File: src/python/liskov_substitution_principle.py
from abc import ABC, abstractmethod


class Bird(ABC):
    def make_sound(self):
        print("Chirp chirp!")


class FlyingBird(Bird):
    def fly(self):
        print("I can fly!")


class Penguin(Bird):
    def swim(self):
        print("I can swim!")


def move_bird(bird: Bird):
    bird.make_sound()


def let_bird_fly(bird: Bird):
    if isinstance(bird, FlyingBird):
        bird.fly()
    else:
        print("This bird can't fly.")


# Usage
move_bird(FlyingBird())  # Chirp chirp!
move_bird(Penguin())  # Chirp chirp!
let_bird_fly(FlyingBird())  # I can fly!
let_bird_fly(Penguin())  # This bird can't fly.

File: src/kotlin/LiskovSubstitutionPrinciple.kt
package kotlin

open class Bird {
    fun makeSound() {
        println("Chirp chirp!")
    }
}

class FlyingBird : Bird() {
    fun fly() {
        println("I can fly!")
    }
}

class Penguin : Bird() {
    fun swim() {
        println("I can swim!")
    }
}

fun moveBird(bird: Bird) {
    bird.makeSound()
}

fun letBirdFly(bird: Bird) {
    if (bird is FlyingBird) {
        bird.fly()
    } else {
        println("This bird can't fly.")
    }
}

// Usage
fun main() {
    moveBird(FlyingBird())   // Chirp chirp!
    moveBird(Penguin())      // Chirp chirp!
    letBirdFly(FlyingBird()) // I can fly!
    letBirdFly(Penguin())    // This bird can't fly.
}
